keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+administrator to configure a cluster where some nodes have this hardware.
+
+I develop two components, one that runs as part of the Kubernetes control plane
+and one that runs on each node, and package those inside container images. YAML
+files describe how to deploy my software on a Kubernetes cluster that supports
+dynamic resource allocation.
+
+Documentation for administrators explains how the nodes need to be set
+up. Documentation for users explains which parameters control the behavior of
+my hardware and how to use it inside a container.
+
+#### Cluster configuration
+
+As a cluster administrator, I want to make GPUs from vendor ACME available to users
+of that cluster. I prepare the nodes and deploy the vendor's components with
+`kubectl create`.
+
+I create a ResourceClass for the hardware with parameters that only I as the
+administrator am allowed to choose, like for example running a command with
+root privileges that does some cluster-specific initialization for each allocation:
+```
+apiVersion: gpu.acme.com/v1
+kind: GPUInit
+metadata:
+  name: acme-gpu-init
+initCommand:
+- /usr/local/bin/acme-gpu-init
+- --cluster
+- my-cluster
+---
+apiVersion: cdi.k8s.io/v1alpha1
+metadata:
+  name: acme-gpu
+driverName: gpu.acme.com
+parameters:
+  apiVersion: gpu.acme.com/v1
+  kind: GPUInit
+  name: acme-gpu-init
+```
+
+#### Partial GPU allocation
+
+As a user, I want to use a GPU as accelerator, but don't need exclusive access
+to that GPU. Running my workload with just 2Gb of memory is sufficient. This is
+supported by the ACME GPU hardware. I know that the administrator has created
+an "acme-gpu" ResourceClass.
+
+For a simple trial, I create a Pod directly where two containers share the same subset
+of the GPU:
+```
+apiVersion: gpu.acme.com/v1
+kind: GPURequirements
+metadata:
+  name: device-consumer-gpu-parameters
+memory: "2Gi"
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: device-consumer
+spec:
+  resourceClaims:
+  - name: "gpu" # this name gets referenced below under "claims"
+    template:
+      resourceClassName: "acme-gpu"
+      parameters:
+        apiVersion: gpu.acme.com/v1
+        kind: GPURequirements
+        name: device-consumer-gpu-parameters
+  containers:
+  - name: workload
+    image: my-app
+    command: ["/bin/program"]
+    resources:
+      requests:
+        memory: "64Mi"
+        cpu: "250m"
+      limits:
+        memory: "128Mi"
+        cpu: "500m"
+      claims:
+        - "gpu"
+  - name: monitor
+    image: my-app
+    command: ["/bin/other-program"]
+    resources:
+      requests:
+        memory: "32Mi"
+        cpu: "25m"
+      limits:
+        memory: "64Mi"
+        cpu: "50m"
+      claims:
+      - "gpu"
+```
+
+This request triggers resource allocation on a node that has a GPU device with
+2Gi of memory available and then the Pod runs on that node. The remaining
+capacity of the GPU can be used by other pods. The lifecycle of the resource
```

Comment:
With "bad segmentation" do you mean additional constraints for the hardware regarding how it can split up memory? Those wouldn't be controlled by the admin, but rather by the resource driver.

In fact, this is one of the key arguments for this KEP: when advertising GPU memory as extended resource, the scheduler knows nothing about such constraints and will split it up arbitrarily among pods. With this KEP, the resource driver can check arbitrary constraints before allocating from a specific GPU that is already in use.

Changed:

> The remaining
capacity of the GPU may be usable for other pods, with constrains like alignment to segment sizes ensured by the resource driver.

Reply:
I'm saying: The driver thinks it can share a resource, by my security team knows they are full of hot air and won't allow it.  I guess that becomes a feature-request of the driver to allow disabling?

Reply:
Yes. Or don't install the resource driver because it made questionable design choices.